DATACASS-812 - Added schema generation support for static columns.

alex-ninja opened the pull request:
Closes #978

- [x] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/master/CONTRIBUTING.adoc).
- [x] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/master/etc/ide) and have them applied to your changes. Don’t submit any formatting related changes.
- [x] You submit test cases (unit or integration tests) that back your changes.
- [x] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).

### Overview of the changes
This PR addresses #978 / DATACASS-812 feature request. The main goal is to support single-value (static) columns. Instead of the suggested in the original issue approach based on an additional attribute in `@Column` annotation (e.g. `@Column(name = "x", isStatic = true)`, I have added a separate `@StaticColumn` annotation. It seems to be more flexible because the `@Column` annotation requires to set a column name (which might be inconvenient when an end user does not want to override the default naming).

alex-ninja commented:
Hi @mp911de !
Could you please briefly take a look and provide an initial feedback on this PR.

mp911de commented on spring-data-cassandra/src/main/java/org/springframework/data/cassandra/core/mapping/StaticColumn.java:
```diff
@@ -0,0 +1,34 @@
+@Documented
+@Retention(RetentionPolicy.RUNTIME)
+@Target(value = { ElementType.ANNOTATION_TYPE, ElementType.FIELD, ElementType.METHOD })
+public @interface StaticColumn {
```
Instead of introducing another annotation type, it seems that we should rather extend `@Column` and probably `@PrimaryKeyColumn` by adding a `staticColumn` or `isStatic` attribute.

alex-ninja replied:
As proposed, I've added `isStatic` attribute to `@Column` annotation. There is no need to add that to `@PrimaryKeyColumn`  annotation because primary columns (partitioned/clustered) cannot be static ([reference to documentation](https://docs.datastax.com/en/dse/6.8/cql/cql/cql_using/refStaticCol.html)). There is a separate check for that:
```
Assert.isTrue(!(optionalKeyType.isPresent() && isStatic),"PrimaryKey must not be static");
```

mp911de commented on spring-data-cassandra/src/main/java/org/springframework/data/cassandra/core/cql/keyspace/ColumnSpecification.java:
```diff
@@ -178,6 +180,18 @@ public ColumnSpecification ordering(Ordering ordering) {
+	/**
+	 * Identifies this column as a static column. Sets the column's {@link #isStatic} to {@literal true}.
+	 *
+	 * @return this
```
Nit: `@since 3.2`

alex-ninja replied:
Done

mp911de commented on spring-data-cassandra/src/main/java/org/springframework/data/cassandra/core/cql/keyspace/TableSpecification.java:
```diff
@@ -82,7 +87,29 @@ public T column(String name, DataType type) {
+	 *          must not be {@literal null}.
+	 * @param type The data type of the column, must not be {@literal null}.
+	 */
+	public T staticColumn(String name, DataType type) {
```
Nit: `@since 3.2`

alex-ninja replied:
Done

mp911de commented on spring-data-cassandra/src/main/java/org/springframework/data/cassandra/core/cql/keyspace/TableSpecification.java:
```diff
@@ -82,7 +87,29 @@ public T column(String name, DataType type) {
+	 *          must not be {@literal null}.
+	 * @param type The data type of the column, must not be {@literal null}.
+	 */
+	public T staticColumn(CqlIdentifier name, DataType type) {
```
Nit: `@since 3.2`

alex-ninja replied:
Done

mp911de commented on spring-data-cassandra/src/main/java/org/springframework/data/cassandra/core/cql/keyspace/TableSpecification.java:
```diff
@@ -317,4 +350,12 @@ protected T column(CqlIdentifier name, DataType type, Optional<PrimaryKeyType> o
+	 * Returns an unmodifiable list of static columns.
+	 */
+	@Override
+	public List<ColumnSpecification> getStaticColumns() {
```
Nit: `@since 3.2`

alex-ninja replied:
Done

mp911de commented on spring-data-cassandra/src/main/java/org/springframework/data/cassandra/core/mapping/CassandraPersistentProperty.java:
```diff
@@ -179,4 +179,13 @@ default boolean isEmbedded() {
+	 * Whether the property is a static column.
+	 * @since 3.2
+	 */
+	default boolean isStaticColumn() {
```
Annotation lookup can be somewhat expensive. It makes sense to implement the lookup directly in `BasicCassandraPersistentProperty` and update `CachingCassandraPersistentProperty` to cache the outcome whether that's a static column.

alex-ninja replied:
Good catch! I've made the corresponding fix and updated `BasicCassandraPersistentTupleProperty` with `false` since it cannot be static. Additionally, I've added a separate test for a static tuple field.

mp911de reviewed (changes requested):
Thanks for your pull request. I left a few comments. Design-wise, we should rather update existing annotations instead of introducing new ones. It aligns better with the purpose of annotations defining their purpose and by doing so we can exclude static columns from being used with `@PrimaryKey`. Since I'm not sure whether static columns make sense in the context of clustering keys, I'm not sure whether we should extend `@PrimaryKeyColumn` as well.

alex-ninja commented:
@mp911de Thanks for valuable feedback! I've addressed all the comments and squashed everything into a single commit (as stated in the contribution guide).

alex-ninja commented:
@mp911de 
Were you able to check the PR? I'm sorry for being slightly impatient, just want to confirm it is not missing from your radar.

mp911de commented:
No need to worry. This PR is still being considered for the upcoming GA release.

mp911de commented:
Thank you for your contribution. That's merged and polished now.
